Objects/typeobject.c

```diff
@@ -1189,8 +1186,10 @@ subtype_dealloc(PyObject *self)
         assert(basedealloc);
         basedealloc(self);
 
-        /* Can't reference self beyond this point */
-        Py_DECREF(type);
+       /* Only decref if the base type is not already a heap allocated type.
+          Otherwise, basedealloc should have decref'd it already */
+        if (type->tp_flags & Py_TPFLAGS_HEAPTYPE && !(base->tp_flags & Py_TPFLAGS_HEAPTYPE))
```

Comment:
Why the first condition `type->tp_flags & Py_TPFLAGS_HEAPTYPE`? There is already an assertion at the beginning of `subtype_dealloc`.

Reply:
The type can be changed by `tp_del`. If you look closely at the implementation, the type is extracted again. So the type that you are looking at this spot might not be the same type that you started off with.

Reply:
> The type can be changed by `tp_del`

I see. Reading the code for `__class__` assignment, this would only be possible for subtypes of `module`. Do you know if there is a testcase for this?

Reply:
It's possible for both subtypes of module and heap types. The specific check done by `object_set_class` in `typeobject.c` is: https://github.com/python/cpython/blob/master/Objects/typeobject.c#L4049-L4057

And yes, I've added two tests for this. One for the `!PyType_IS_GC` case (which is the one you have here). And for the `PyType_IS_GC` case down below.

Both tests have been added to `test_capi.py` and are named `test_c_subclass_of_heap_ctype_with_del_modifying_dunder_class_only_decrefs_once` and `test_subclass_of_heap_gc_ctype_with_del_modifying_dunder_class_only_decrefs_once` respectively.